Index: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelProcessor.java
===================================================================
--- maven-model-builder/src/main/java/org/apache/maven/model/building/ModelProcessor.java	(revision 824608)
+++ maven-model-builder/src/main/java/org/apache/maven/model/building/ModelProcessor.java	(working copy)
@@ -6,4 +6,5 @@
 public interface ModelProcessor
     extends ModelLocator, ModelReader
 {
+    String LOCATION = "org.apache.maven.model.building.location";
 }
Index: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java
===================================================================
--- maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java	(revision 824608)
+++ maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java	(working copy)
@@ -28,6 +28,7 @@
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.HashMap;
 
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.DependencyManagement;
@@ -298,7 +299,9 @@
         {
             boolean strict = request.getValidationLevel() >= ModelBuildingRequest.VALIDATION_LEVEL_MAVEN_2_0;
 
-            Map<String, ?> options = Collections.singletonMap( ModelProcessor.IS_STRICT, Boolean.valueOf( strict ) );
+            Map<String,Object> options = new HashMap<String,Object>();
+            options.put(ModelProcessor.IS_STRICT, Boolean.valueOf( strict ));
+            options.put(ModelProcessor.LOCATION, modelSource.getLocation());
 
             model = modelProcessor.read( modelSource.getInputStream(), options );
         }
